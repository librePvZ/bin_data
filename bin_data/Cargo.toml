[package]
name = "bin_data"
version = "0.1.0"
edition = "2021"
description = "declarative encoding and decoding for binary formats"

authors.workspace = true
repository.workspace = true
homepage.workspace = true
license.workspace = true
readme.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
macros = ["dep:bin_data_macros"]

[dependencies]
thiserror = "1.0.40"
bin_data_macros = { path = "../bin_data_macros", optional = true }

[dev-dependencies]
bin_data_macros = { path = "../bin_data_macros" }
