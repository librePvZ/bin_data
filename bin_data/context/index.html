<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Context for encoding and decoding."><meta name="keywords" content="rust, rustlang, rust-lang, context"><title>bin_data::context - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-1f7d512b176f0f72.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Bold-124a1ca42af929b6.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-93196c7a1c3542a8.css" id="mainThemeStyle"><link rel="stylesheet" id="themeStyle" href="../../static.files/light-4743e13df3dfe8c4.css"><link rel="stylesheet" disabled href="../../static.files/dark-0e1b889528bd466b.css"><link rel="stylesheet" disabled href="../../static.files/ayu-65289d5d067c7c66.css"><script id="default-settings" ></script><script src="../../static.files/storage-d43fa987303ecbbb.js"></script><script defer src="../../static.files/main-3367e395607fafc1.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../../bin_data/index.html"><img class="rust-logo" src="../../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../../bin_data/index.html"><img class="rust-logo" src="../../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Module context</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-5ec35bf9ca753509.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">bin_data</a>::<wbr><a class="mod" href="#">context</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../../src/bin_data/context.rs.html#1-222">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Context for encoding and decoding.</p>
<p>The types can be roughly grouped into two categories:</p>
<ul>
<li><a href="trait.Context.html#associatedtype.EndianContext" title="Context::EndianContext"><code>Context::EndianContext</code></a>: explicit endianness specification and inheritance.
<ul>
<li><a href="enum.Endian.html" title="Endian"><code>Endian</code></a>: little-endian or big-endian.</li>
<li><a href="struct.NoEndian.html" title="NoEndian"><code>NoEndian</code></a>: endianness is not decided at runtime.</li>
</ul>
</li>
<li><a href="trait.Context.html#associatedtype.ArgsBuilder" title="Context::ArgsBuilder"><code>Context::ArgsBuilder</code></a>: type-level construct for named arguments.
<ul>
<li><a href="struct.NoArgs.html" title="NoArgs"><code>NoArgs</code></a>: no argument at all, or <code>Args = ()</code>.</li>
<li><a href="struct.VecArgs.html" title="VecArgs"><code>VecArgs</code></a> and <a href="struct.VecArgsBuilder.html" title="VecArgsBuilder"><code>VecArgsBuilder</code></a>: arguments for <a href="https://doc.rust-lang.org/1.68.0/alloc/vec/struct.Vec.html" title="Vec"><code>Vec</code></a>, <a href="https://doc.rust-lang.org/1.68.0/std/primitive.slice.html" title="slice"><code>slice</code></a>s, etc.</li>
<li><a href="struct.StrArgs.html" title="StrArgs"><code>StrArgs</code></a> and <a href="struct.StrArgsBuilder.html" title="StrArgsBuilder"><code>StrArgsBuilder</code></a>: arguments for <a href="https://doc.rust-lang.org/1.68.0/alloc/string/struct.String.html" title="String"><code>String</code></a>, <a href="https://doc.rust-lang.org/1.68.0/std/primitive.str.html" title="str"><code>str</code></a>, etc.</li>
</ul>
</li>
</ul>
<p>Types in this module might appear in error messages, here is an overview:</p>
<ul>
<li><strong>expected enum <a href="enum.Endian.html" title="Endian"><code>Endian</code></a>, found struct <a href="struct.NoEndian.html" title="NoEndian"><code>NoEndian</code></a></strong>: endianness for one of the fields
must be specified. Add the following attribute: <code>#[bin_data(endian = &quot;...&quot;)]</code>, either to
the whole <code>struct</code>, or to individual fields.</li>
<li><strong>expected struct <a href="struct.NoEndian.html" title="NoEndian"><code>NoEndian</code></a>, found enum <a href="enum.Endian.html" title="Endian"><code>Endian</code></a></strong>: opposite situation to the previous
one, one of the fields require no runtime-specified endianness, but endianness is specified
explicitly using <code>#[bin_data(endian = &quot;...&quot;)]</code>. Remove that superfluous attribute.</li>
<li><strong><em>some argument builder</em> does not implement <a href="trait.ArgsBuilderFinished.html" title="ArgsBuilderFinished"><code>ArgsBuilderFinished</code></a></strong>: some <a href="struct.Required.html" title="Required"><code>Required</code></a>
argument is not specified. Specify it in <code>#[bin_data(args { ... })]</code>. The arguments should
be given as <code>name = value</code>, separated and optionally ended by commas. Under the hood, it
calls the method <code>name</code> with argument <code>value</code> on the argument builder. For instance, the
following code set the expected length of a <a href="https://doc.rust-lang.org/1.68.0/alloc/vec/struct.Vec.html" title="Vec"><code>Vec</code></a> by calling <a href="struct.VecArgsBuilder.html#method.count" title="VecArgsBuilder::count"><code>VecArgsBuilder::count</code></a>:
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[bin_data(args:decode { count = <span class="number">42 </span>})]
</span>xs: Vec&lt;u8&gt;,</code></pre></div>
</li>
</ul>
<p>See also each type’s documentation for detailed explanation.</p>
</div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.NoArgs.html" title="bin_data::context::NoArgs struct">NoArgs</a></div><div class="item-right docblock-short">Trivial arguments builder for <code>()</code>.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.NoEndian.html" title="bin_data::context::NoEndian struct">NoEndian</a></div><div class="item-right docblock-short">Indicate the endianness is not determined at runtime.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Provided.html" title="bin_data::context::Provided struct">Provided</a></div><div class="item-right docblock-short">A provided field. This is deliberately not <a href="https://doc.rust-lang.org/1.68.0/core/default/trait.Default.html" title="Default"><code>Default</code></a>, to prevent accidentally supplying
default values for required arguments.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Required.html" title="bin_data::context::Required struct">Required</a></div><div class="item-right docblock-short">Placeholder for a required field without a default value.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.StrArgs.html" title="bin_data::context::StrArgs struct">StrArgs</a></div><div class="item-right docblock-short">Arguments for encoding or decoding a <a href="https://doc.rust-lang.org/1.68.0/std/primitive.str.html" title="str"><code>str</code></a>, <a href="https://doc.rust-lang.org/1.68.0/alloc/string/struct.String.html" title="String"><code>String</code></a>, etc.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.StrArgsBuilder.html" title="bin_data::context::StrArgsBuilder struct">StrArgsBuilder</a></div><div class="item-right docblock-short">Named arguments builder for <a href="struct.StrArgs.html" title="StrArgs"><code>StrArgs</code></a>.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.VecArgs.html" title="bin_data::context::VecArgs struct">VecArgs</a></div><div class="item-right docblock-short">Arguments for encoding or decoding a <a href="https://doc.rust-lang.org/1.68.0/alloc/vec/struct.Vec.html" title="Vec"><code>Vec</code></a>.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.VecArgsBuilder.html" title="bin_data::context::VecArgsBuilder struct">VecArgsBuilder</a></div><div class="item-right docblock-short">Named arguments builder for <a href="struct.VecArgs.html" title="VecArgs"><code>VecArgs</code></a>.</div></div></div><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Endian.html" title="bin_data::context::Endian enum">Endian</a></div><div class="item-right docblock-short">Endianness for integers, floating-point numbers, etc.</div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.ArgsBuilderFinished.html" title="bin_data::context::ArgsBuilderFinished trait">ArgsBuilderFinished</a></div><div class="item-right docblock-short">Indicates that all arguments is supplied.</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Context.html" title="bin_data::context::Context trait">Context</a></div><div class="item-right docblock-short">Specify the named argument builder and endianness context.</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="bin_data" data-themes="" data-resource-suffix="" data-rustdoc-version="1.68.0 (2c8cc3432 2023-03-06)" data-search-js="search-98d53477a794af0b.js" data-settings-js="settings-c3c521c753752a1a.js" data-settings-css="settings-08ddfdda51b8ee2e.css" ></div></body></html>