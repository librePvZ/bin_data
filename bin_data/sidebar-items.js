window.SIDEBAR_ITEMS = {"macro":[["bin_data","Declare a binary data format."]],"mod":[["context","Context for encoding and decoding."],["data","Interface for encoding and decoding binary data."],["stream","Input and output streams for binary data."]]};