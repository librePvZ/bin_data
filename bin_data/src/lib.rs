//! Declarative encoding and decoding for binary formats.

#![warn(missing_docs)]
#![warn(missing_debug_implementations)]

pub mod context;
pub mod stream;
pub mod data;
