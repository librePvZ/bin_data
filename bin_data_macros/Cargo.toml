[package]
name = "bin_data_macros"
version = "0.1.0"
edition = "2021"
description = "supporting macros for bin_data"

authors.workspace = true
repository.workspace = true
homepage.workspace = true
license.workspace = true
readme.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
itertools = "0.10.5"
proc-macro2 = "1.0.52"
quote = "1.0.26"
syn = "2.0.0"

[dev-dependencies]
bin_data.path = "../bin_data"
