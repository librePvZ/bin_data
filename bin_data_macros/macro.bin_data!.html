<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.bin_data.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.bin_data.html">macro.bin_data.html</a>...</p>
    <script>location.replace("macro.bin_data.html" + location.search + location.hash);</script>
</body>
</html>