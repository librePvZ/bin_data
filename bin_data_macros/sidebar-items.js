window.SIDEBAR_ITEMS = {"macro":[["bin_data","Declare a binary data format."]]};