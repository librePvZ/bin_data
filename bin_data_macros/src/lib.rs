//! Supporting macros for `bin_data`.

#![warn(missing_docs)]
