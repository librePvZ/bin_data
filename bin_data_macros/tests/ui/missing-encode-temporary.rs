use bin_data_macros::bin_data;

bin_data! {
    #[bin_data(endian = "little")]
    pub struct MissingEncode {
        let temporary: u32,
        #[bin_data(decode = temporary)]
        pub field: u32,
    }
}

fn main() {}
