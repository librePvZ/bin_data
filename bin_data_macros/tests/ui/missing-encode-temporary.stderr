error: temporary field requires an `encode` attribute
 --> tests/ui/missing-encode-temporary.rs:6:13
  |
6 |         let temporary: u32,
  |             ^^^^^^^^^
